### ISSUE TITLE
PBS: SKAdNetwork imp.ext passthrough

### ISSUE THREAD

**Issue by bszekely1** [prebid/prebid-server#1654]
## Type of issue

As part of support for [SKAdNetwork](https://github.com/prebid/prebid-mobile-ios/issues/342), Prebid SDK requires passing an object to PBS Bid Adapters in the `imp.ext.skadn` object. The issue here is that PBS is treating the `skadn` object as a bidder and failing on validation, since `skadn` is not a bidder due to https://github.com/prebid/prebid-server/issues/1478.

## Requirements
The requirements of this ticket are the following:
- Ensure `imp.ext.skadn` is treated as a non-bider valid key
- Ensure `imp.ext.skadn` is passed to Bid Adapters
- Ensure the response object `seatbid.bid.ext.skadn` received from bidders is forwarded back to Prebid SDK

sample skadn request object:

```
{
  "imp": [
    {
      "ext": {
        "skadn": {
          "version": "2.0",
          "sourceapp": "880047117",
          "skadnetlist":{
              "max":306,
              "excl":[2,8,10,55],
              "addl":[
                "cDkw7geQsH.skadnetwork",
                "qyJfv329m4.skadnetwork"
              ]
          }
        }
      }
    }
  ]
}
```

sample skadn response object:
```
{
  "seatbid": [
    {
      "bid": [
        {
          "ext": {
            "skadn": {
              "version": "2.0",
              "network": "cDkw7geQsH.skadnetwork",
              "campaign": "45",
              "itunesitem": "123456789",
              "nonce": "473b1a16-b4ef-43ad-9591-fcf3aefa82a7",
              "sourceapp": "880047117",
              "timestamp": "1594406341",
              "signature": "MEQCIEQlmZRNfYzKBSE8QnhLTIHZZZWCFgZpRqRxHss65KoFAiAJgJKjdrWdkLUOCCjuEx2RmFS7daRzSVZRVZ8RyMyUXg=="
            }
          }
        }
      ]
    }
  ]
}
```

## Other information

https://github.com/prebid/prebid-mobile-ios/issues/342

**Comment by SyntaxNode** [comment]
@bszekely1 To clarify "Ensure the response object seatbid.bid.ext.skadn is forwarded back to Prebid SDK", should PBS Core copy `imp.ext.skadn` from the request to the `seatbid.bid.ext.skadn` of each imp bid... or is this an ask to allow a bid adapter to set it (which I don't think is possible now)?

**Comment by bszekely1** [comment]
> @bszekely1 To clarify "Ensure the response object seatbid.bid.ext.skadn is forwarded back to Prebid SDK", should PBS Core copy `imp.ext.skadn` from the request to the `seatbid.bid.ext.skadn` of each imp bid... or is this an ask to allow a bid adapter to set it (which I don't think is possible now)?

I added a little more clarity on the ticket. The response comes from bid adapters. Anything in the `seatbid.bid.ext.skadn` object needs to be received by Prebid SDK.

**Comment by SyntaxNode** [comment]
Got it. Thank you for the clarification.

PBS doesn't currently allow the bidder to wrote ext values at the top level. It pushes everything into `seatbid.bid.ext.bidder`. Could the Prebid SDK support reading the `skadn` data from `seatbid.bid.ext.bidder.skadn` or does it really need to be `seatbid.bid.ext.skadn`?

Also, should PBS perform validation on the `skadn` json or treat is as opaque data?

**Comment by bszekely1** [comment]
This is an OpenRTB thing, not a Prebid thing. Technically speaking there is no hard requirement to keep it as OpenRTB between Prebid SDK and PBS, but I personally do not like the idea of moving bidder bid.ext objects into a prebid object. This is a topic we will be speaking about tomorrow on the PBS PMC call.

**Comment by SyntaxNode** [comment]
> This is an OpenRTB thing, not a Prebid thing

Ah, alright. We can add a new field to the adapter bid response which lets them set the `skadn` value for the core to write to `seatbid.bid.ext.skadn`. That should work well with the pattern we have today for the `seatbid.bid.ext.bidder` and `seatbid.bid.ext.prebid` sections.

Let's discuss tomorrow if PBS should be performing validation on the data type or just facilitate passing through json blobs.

**Comment by bretg** [comment]
Discussed in PBS Committee.

- Proposal is that PBS will experiment with dropping the behavior of moving bidder ext to seatbid.bid.ext.bidder. We are not aware of anything that looks at seatbid.bid.ext.bidder: SDK, AMP, or PBJS. We will communicate with the adapter community to see if anyone is relying on this field.
- This means no exception would be needed for seatbid.bid.ext.skadn.
- We propose doing no validations on the attribute's value. It's unclear to us what value PBS would bring by policing this attribute, and the possibility of future changes makes us question the utility of validations.
- imp.ext.skadn on the input side will be made an exception

**Comment by SyntaxNode** [comment]
We decided:
- PBS will treat the skadn json object as an opaque identifier. It will be passed down to adapters, but not validated.
- PBS will make a breaking change where bid exts will be placed directly in `seatbid.bid.ext` instead of in `seatbid.bid.ext.bidder`. I'll open a separate issue to track that.

**Comment by SyntaxNode** [comment]
Created https://github.com/prebid/prebid-server/issues/1676 for the response side. Let's consider this story complete when the request side exception is complete.

**Comment by bretg** [comment]
Done in PBS-Java 1.58

**Comment by SyntaxNode** [comment]
Implemented in PBS-Go 0.151.0
